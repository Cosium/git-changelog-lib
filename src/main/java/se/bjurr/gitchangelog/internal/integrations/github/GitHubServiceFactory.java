package se.bjurr.gitchangelog.internal.integrations.github;

import static java.util.concurrent.TimeUnit.SECONDS;

import com.google.common.base.Optional;
import java.io.File;
import java.io.IOException;
import okhttp3.Cache;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class GitHubServiceFactory {

  private static GitHubService gitHubService = null;

  public static void reset() {
    gitHubService = null;
  }

  public static GitHubService getGitHubService(String api, final Optional<String> token) {
    return getGitHubService(api, token, null);
  }

  public static synchronized GitHubService getGitHubService(
      String api, final Optional<String> token, final Interceptor interceptor) {
    if (!api.endsWith("/")) {
      api += "/";
    }
    if (gitHubService == null) {
      File cacheDir = new File(".okhttpcache");
      cacheDir.mkdir();
      Cache cache = new Cache(cacheDir, 1024 * 1024 * 10);

      OkHttpClient.Builder builder =
          new OkHttpClient.Builder().cache(cache).connectTimeout(10, SECONDS);

      if (token != null && token.isPresent() && !token.get().isEmpty()) {
        builder.addInterceptor(
            new Interceptor() {
              @Override
              public Response intercept(Chain chain) throws IOException {
                Request original = chain.request();

                Request request =
                    original
                        .newBuilder() //
                        .addHeader("Authorization", "token " + token.get()) //
                        .method(original.method(), original.body()) //
                        .build();
                return chain.proceed(request);
              }
            });
      }

      if (interceptor != null) {
        builder.addInterceptor(interceptor);
      }

      Retrofit retrofit =
          new Retrofit.Builder() //
              .baseUrl(api) //
              .client(builder.build()) //
              .addConverterFactory(GsonConverterFactory.create()) //
              .build();

      gitHubService = retrofit.create(GitHubService.class);
    }
    return gitHubService;
  }
}
