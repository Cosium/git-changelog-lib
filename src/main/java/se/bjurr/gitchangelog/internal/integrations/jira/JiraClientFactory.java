package se.bjurr.gitchangelog.internal.integrations.jira;

public class JiraClientFactory {

 private static JiraClient jiraClient;

 public static void reset() {
  jiraClient = null;
 }

 public static JiraClient createJiraClient(String apiUrl) {
  if (jiraClient == null) {
   jiraClient = new JiraClient(apiUrl);
  }
  return jiraClient;
 }

}
