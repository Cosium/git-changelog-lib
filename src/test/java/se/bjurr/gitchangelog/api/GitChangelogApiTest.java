package se.bjurr.gitchangelog.api;

import static com.google.common.collect.ImmutableMap.of;
import static com.google.common.collect.Maps.newHashMap;
import static com.google.common.io.Resources.getResource;
import static java.nio.charset.StandardCharsets.UTF_8;
import static org.assertj.core.api.Assertions.assertThat;
import static se.bjurr.gitchangelog.api.GitChangelogApi.gitChangelogApiBuilder;
import static se.bjurr.gitchangelog.api.GitChangelogApiConstants.ZERO_COMMIT;
import static se.bjurr.gitchangelog.internal.integrations.rest.RestClient.mock;

import com.google.common.io.Resources;
import java.net.URL;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import se.bjurr.gitchangelog.internal.integrations.github.GitHubMockInterceptor;
import se.bjurr.gitchangelog.internal.integrations.github.GitHubServiceFactory;
import se.bjurr.gitchangelog.internal.integrations.jira.JiraClientFactory;
import se.bjurr.gitchangelog.internal.integrations.rest.RestClientMock;
import se.bjurr.gitchangelog.test.ApprovalsWrapper;

public class GitChangelogApiTest {
  private RestClientMock mockedRestClient;
  private GitHubMockInterceptor gitHubMockInterceptor;

  @Before
  public void before() throws Exception {
    JiraClientFactory.reset();

    this.mockedRestClient = new RestClientMock();
    this.mockedRestClient //
        .addMockedResponse(
            "/repos/tomasbjerre/git-changelog-lib/issues?state=all",
            Resources.toString(getResource("github-issues.json"), UTF_8)) //
        .addMockedResponse(
            "/jira/rest/api/2/issue/JIR-1234?fields=parent,summary,issuetype,labels,description,issuelinks",
            Resources.toString(getResource("jira-issue-jir-1234.json"), UTF_8)) //
        .addMockedResponse(
            "/jira/rest/api/2/issue/JIR-5262?fields=parent,summary,issuetype,labels,description,issuelinks",
            Resources.toString(getResource("jira-issue-jir-5262.json"), UTF_8));
    mock(this.mockedRestClient);

    this.gitHubMockInterceptor = new GitHubMockInterceptor();
    this.gitHubMockInterceptor //
        .addMockedResponse(
        "https://api.github.com/repos/tomasbjerre/git-changelog-lib/issues?state=all&per_page=100&page=1",
        Resources.toString(getResource("github-issues.json"), UTF_8));

    GitHubServiceFactory //
        .setInterceptor(this.gitHubMockInterceptor);
  }

  @After
  public void after() {
    JiraClientFactory.reset();
    GitHubServiceFactory.setInterceptor(null);
    mock(null);
  }

  @Test
  public void testThatTagsThatAreEmptyAfterCommitsHaveBeenIgnoredAreRemoved() throws Exception {
    final String templatePath = "templatetest/testAuthorsCommitsExtended.mustache";

    final GitChangelogApi given =
        gitChangelogApiBuilder() //
            .withFromCommit(ZERO_COMMIT) //
            .withToRef("test") //
            .withTemplatePath(templatePath) //
            .withIgnoreCommitsWithMessage(".*");

    ApprovalsWrapper.verify(given);
  }

  @Test
  public void testPathFilterCanBeSpecified() throws Exception {
    final String templatePath = "templatetest/testAuthorsCommitsExtended.mustache";

    final GitChangelogApi given =
        gitChangelogApiBuilder() //
            .withFromCommit(ZERO_COMMIT) //
            .withToRef("1.71") //
            .withTemplatePath(templatePath) //
            .withPathFilter("src");

    ApprovalsWrapper.verify(given);
  }

  @Test
  public void testThatIssuesCanBeRemoved() throws Exception {

    final URL settingsFile =
        getResource("settings/git-changelog-test-settings.json").toURI().toURL();
    final String templatePath = "templatetest/testIssuesCommits.mustache";

    final GitChangelogApi given =
        gitChangelogApiBuilder() //
            .withFromCommit(ZERO_COMMIT) //
            .withToRef("test") //
            .withSettings(settingsFile) //
            .withRemoveIssueFromMessageArgument(true) //
            .withTemplatePath(templatePath);

    ApprovalsWrapper.verify(given);
  }

  @Test
  public void testThatCommitsWithoutIssueCanBeIgnoredIssuesCommits() throws Exception {

    final URL settingsFile =
        getResource("settings/git-changelog-test-settings.json").toURI().toURL();
    final String templatePath = "templatetest/testIssuesCommits.mustache";

    Resources.toString(getResource(templatePath), UTF_8);

    final GitChangelogApi given =
        gitChangelogApiBuilder() //
            .withFromCommit(ZERO_COMMIT) //
            .withToRef("test") //
            .withSettings(settingsFile) //
            .withIgnoreCommitsWithoutIssue(true) //
            .withTemplatePath(templatePath);

    ApprovalsWrapper.verify(given);
  }

  /** The test-lightweight-2 should be ignored here. */
  @Test
  public void testThatCommitsWithoutIssueCanBeIgnoredTagsIssuesCommits() throws Exception {

    final String templatePath =
        "templatetest/testThatCommitsWithoutIssueCanBeIgnoredTagsIssuesCommits.mustache";

    final GitChangelogApi given =
        gitChangelogApiBuilder() //
            .withFromCommit(ZERO_COMMIT) //
            .withToRef("test") //
            .withGitHubIssuePattern("nooo") //
            .withGitLabIssuePattern("nooo") //
            .withCustomIssue(
                "JIRA", "JIR-[0-9]*", "http://${PATTERN_GROUP}", "${PATTERN_GROUP}") //
            .withIgnoreCommitsWithoutIssue(true) //
            .withTemplatePath(templatePath);

    ApprovalsWrapper.verify(given);
  }

  @Test
  public void testThatReadableGroupMustExist() throws Exception {
    final URL settingsFile =
        getResource("settings/git-changelog-test-settings.json").toURI().toURL();
    final String templatePath = "templatetest/testIssuesCommits.mustache";

    try {
      final String actual =
          gitChangelogApiBuilder() //
              .withFromCommit(ZERO_COMMIT) //
              .withSettings(settingsFile) //
              .withToRef("/test") //
              .withRemoveIssueFromMessageArgument(true) //
              .withTemplatePath(templatePath) //
              .withReadableTagName("[0-9]+?") //
              .render();
      assertThat(actual) //
          .as(
              "Should never happen! But nice to see what was rendered, if it does not crash as expected.") //
          .isEqualTo("");
    } catch (final Exception e) {
      if (!e.getMessage()
          .equals(
              "Pattern: \"[0-9]+?\" did not match any group in: \"refs/tags/test-lightweight-2\"")) {
        throw new AssertionError("Got: \"" + e.getMessage() + "\"", e);
      }
    }
  }

  @Test
  public void testThatReadableGroupCanBeSet() throws Exception {
    final URL settingsFile =
        getResource("settings/git-changelog-test-settings.json").toURI().toURL();
    final String templatePath = "templatetest/testIssuesCommits.mustache";

    final GitChangelogApi given =
        gitChangelogApiBuilder() //
            .withFromCommit(ZERO_COMMIT) //
            .withToRef("test") //
            .withSettings(settingsFile) //
            .withRemoveIssueFromMessageArgument(true) //
            .withTemplatePath(templatePath) //
            .withReadableTagName(".*/([0-9]+?\\.[0-9]+?)$");

    ApprovalsWrapper.verify(given);
  }

  @Test
  public void testThatCustomVariablesCanBeUsed() throws Exception {
    final URL settingsFile =
        getResource("settings/git-changelog-test-settings.json").toURI().toURL();
    final String templatePath = "templatetest/testAuthorsCommitsExtended.mustache";

    final GitChangelogApi given =
        gitChangelogApiBuilder() //
            .withFromCommit(ZERO_COMMIT) //
            .withToRef("test") //
            .withSettings(settingsFile) //
            .withExtendedVariables(newHashMap(of("customVariable", (Object) "the value"))) //
            .withRemoveIssueFromMessageArgument(true) //
            .withTemplatePath(templatePath);

    ApprovalsWrapper.verify(given);
  }

  @Test
  public void shouldUseIntegrationIfConfigured() {
    assertThat(GitChangelogApi.shouldUseIntegrationIfConfigured("just title is false")) //
        .isFalse();

    assertThat(GitChangelogApi.shouldUseIntegrationIfConfigured("{{title}}")) //
        .isTrue();

    assertThat(GitChangelogApi.shouldUseIntegrationIfConfigured("{{link}}")) //
        .isTrue();

    assertThat(GitChangelogApi.shouldUseIntegrationIfConfigured("{{type}}")) //
        .isTrue();

    assertThat(GitChangelogApi.shouldUseIntegrationIfConfigured("asd{{#labels}}")) //
        .isTrue();
    assertThat(GitChangelogApi.shouldUseIntegrationIfConfigured("a\nsd{{#labels}}asd\nsdasd")) //
        .isTrue();
    assertThat(GitChangelogApi.shouldUseIntegrationIfConfigured("a\nsd{{labels}}asd\nsdasd")) //
        .isTrue();
  }

  @Test
  public void testThatRevertedCommitsAreRemoved() throws Exception {
    final String templatePath = "templatetest/testThatRevertedCommitsAreRemoved.mustache";

    final GitChangelogApi given =
        gitChangelogApiBuilder() //
            .withFromCommit("aa1fd33") //
            .withToCommit("4c6e078") //
            .withTemplatePath(templatePath);

    ApprovalsWrapper.verify(given);
  }
}
