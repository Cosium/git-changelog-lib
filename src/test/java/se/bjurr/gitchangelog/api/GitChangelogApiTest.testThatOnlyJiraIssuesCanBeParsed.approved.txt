template:

# {{repoName}} changelog

Changelog of {{repoName}}.

{{#tags}}
{{#ifEquals name "Unreleased"}}{{else}}
## {{name}} ({{tagDate .}})

{{#ifContainsBreaking commits}}
### Breaking changes

{{#commits}}
{{#ifCommitBreaking .}}
- {{#eachCommitScope .}} **{{.}}** {{/eachCommitScope}} {{{commitDescription .}}} ([{{subString hash 0 5}}](https://github.com/{{ownerName}}/{{repoName}}/commit/{{hash}}) {{authorName}}) {{#eachCommitRefs .}}{{#ifMatches . "^#[0-9]+"}} [{{.}}](https://github.com/{{ownerName}}/{{repoName}}/issues/{{subString . 1}}) {{/ifMatches}}{{/eachCommitRefs}} {{#eachCommitFixes .}}{{#ifMatches . "^#[0-9]+"}} [{{.}}](https://github.com/{{ownerName}}/{{repoName}}/issues/{{subString . 1}}) {{/ifMatches}}{{/eachCommitFixes}}
{{/ifCommitBreaking}}
{{/commits}}
{{/ifContainsBreaking}}


{{#ifContainsType commits type='feat'}}
### Features

{{#commits}}
{{#ifCommitType . type='feat'}}
- {{#eachCommitScope .}} **{{.}}** {{/eachCommitScope}} {{{commitDescription .}}} ([{{subString hash 0 5}}](https://github.com/{{ownerName}}/{{repoName}}/commit/{{hash}}) {{authorName}}) {{#eachCommitRefs .}}{{#ifMatches . "^#[0-9]+"}} [{{.}}](https://github.com/{{ownerName}}/{{repoName}}/issues/{{subString . 1}}) {{/ifMatches}}{{/eachCommitRefs}} {{#eachCommitFixes .}}{{#ifMatches . "^#[0-9]+"}} [{{.}}](https://github.com/{{ownerName}}/{{repoName}}/issues/{{subString . 1}}) {{/ifMatches}}{{/eachCommitFixes}}
{{/ifCommitType}}
{{/commits}}
{{/ifContainsType}}


{{#ifContainsType commits type='fix'}}
### Bug Fixes

{{#commits}}
{{#ifCommitType . type='fix'}}
- {{#eachCommitScope .}} **{{.}}** {{/eachCommitScope}} {{{commitDescription .}}} ([{{subString hash 0 5}}](https://github.com/{{ownerName}}/{{repoName}}/commit/{{hash}}) {{authorName}}) {{#eachCommitRefs .}}{{#ifMatches . "^#[0-9]+"}} [{{.}}](https://github.com/{{ownerName}}/{{repoName}}/issues/{{subString . 1}}) {{/ifMatches}}{{/eachCommitRefs}} {{#eachCommitFixes .}}{{#ifMatches . "^#[0-9]+"}} [{{.}}](https://github.com/{{ownerName}}/{{repoName}}/issues/{{subString . 1}}) {{/ifMatches}}{{/eachCommitFixes}}
{{/ifCommitType}}
{{/commits}}
{{/ifContainsType}}


{{#ifContainsType commits type='^($|(?!fix|feat|breaking))'}}
### Other changes

{{#commits}}
{{#ifCommitType . type='^($|(?!fix|feat|breaking))'}}
**{{{messageTitle}}}**

{{#messageBodyItems}}
* {{.}} 
{{/messageBodyItems}}

[{{subString hash 0 5}}](https://github.com/{{ownerName}}/{{repoName}}/commit/{{hash}}) {{authorName}} *{{commitTime}}*

{{/ifCommitType}}
{{/commits}}
{{/ifContainsType}}

{{/ifEquals}}
{{/tags}}

---------------------------------------------

settings:

{
  "toRef": "1.71",
  "fromCommit": "0000000000000000000000000000000000000000",
  "removeIssueFromMessage": false,
  "jiraEnabled": true,
  "redmineEnabled": false,
  "gitHubEnabled": false,
  "gitHubApi": "https://api.github.com/repos/tomasbjerre/git-changelog-lib",
  "extendedVariables": {},
  "ignoreCommitsWithoutIssue": true,
  "gitLabEnabled": false,
  "gitLabProjectName": "tomasbjerre",
  "useIntegrations": false,
  "subDirFilter": "src"
}

---------------------------------------------

changelog:

# git-changelog-lib changelog

Changelog of git-changelog-lib.

## 1.71 (2017-07-25)







### Other changes

**Including correct commits + performance #26 JENKINS-34156**

* Found major performance problem when sorting tags by commit time, fixed. 
* Now not following parents, unless *from* is merged into them. 

[e3106](https://github.com/tomasbjerre/git-changelog-lib/commit/e3106df640b693b) Tomas Bjerre *2016-04-13 18:30:26*

**JENKINS-34155 Support short SHA**


[27642](https://github.com/tomasbjerre/git-changelog-lib/commit/27642e3db66e67c) Tomas Bjerre *2016-04-12 16:43:21*




---------------------------------------------

context:

{
  "commits": [
    {
      "authorEmailAddress": "tomas.bjerre85@gmail.com",
      "authorName": "Tomas Bjerre",
      "commitTime": "2016-04-13 18:30:26",
      "commitTimeLong": 1460572226000,
      "hash": "e3106df640b693b",
      "hashFull": "e3106df640b693b50c297154919cb5564895c08e",
      "merge": false,
      "message": "Including correct commits + performance #26 JENKINS-34156\n\n * Found major performance problem when sorting tags by commit time, fixed.\n * Now not following parents, unless *from* is merged into them."
    },
    {
      "authorEmailAddress": "tomas.bjerre85@gmail.com",
      "authorName": "Tomas Bjerre",
      "commitTime": "2016-04-12 16:43:21",
      "commitTimeLong": 1460479401000,
      "hash": "27642e3db66e67c",
      "hashFull": "27642e3db66e67cfe17cdbcc2d6dbf8a3bbcf644",
      "merge": false,
      "message": "JENKINS-34155 Support short SHA"
    }
  ],
  "tags": [
    {
      "annotation": "[Gradle Release Plugin] - creating tag:  \u00271.71\u0027.\n",
      "authors": [
        {
          "commits": [
            {
              "authorEmailAddress": "tomas.bjerre85@gmail.com",
              "authorName": "Tomas Bjerre",
              "commitTime": "2016-04-13 18:30:26",
              "commitTimeLong": 1460572226000,
              "hash": "e3106df640b693b",
              "hashFull": "e3106df640b693b50c297154919cb5564895c08e",
              "merge": false,
              "message": "Including correct commits + performance #26 JENKINS-34156\n\n * Found major performance problem when sorting tags by commit time, fixed.\n * Now not following parents, unless *from* is merged into them."
            },
            {
              "authorEmailAddress": "tomas.bjerre85@gmail.com",
              "authorName": "Tomas Bjerre",
              "commitTime": "2016-04-12 16:43:21",
              "commitTimeLong": 1460479401000,
              "hash": "27642e3db66e67c",
              "hashFull": "27642e3db66e67cfe17cdbcc2d6dbf8a3bbcf644",
              "merge": false,
              "message": "JENKINS-34155 Support short SHA"
            }
          ],
          "authorName": "Tomas Bjerre",
          "authorEmail": "tomas.bjerre85@gmail.com"
        }
      ],
      "commits": [
        {
          "authorEmailAddress": "tomas.bjerre85@gmail.com",
          "authorName": "Tomas Bjerre",
          "commitTime": "2016-04-13 18:30:26",
          "commitTimeLong": 1460572226000,
          "hash": "e3106df640b693b",
          "hashFull": "e3106df640b693b50c297154919cb5564895c08e",
          "merge": false,
          "message": "Including correct commits + performance #26 JENKINS-34156\n\n * Found major performance problem when sorting tags by commit time, fixed.\n * Now not following parents, unless *from* is merged into them."
        },
        {
          "authorEmailAddress": "tomas.bjerre85@gmail.com",
          "authorName": "Tomas Bjerre",
          "commitTime": "2016-04-12 16:43:21",
          "commitTimeLong": 1460479401000,
          "hash": "27642e3db66e67c",
          "hashFull": "27642e3db66e67cfe17cdbcc2d6dbf8a3bbcf644",
          "merge": false,
          "message": "JENKINS-34155 Support short SHA"
        }
      ],
      "issues": [
        {
          "commits": [
            {
              "authorEmailAddress": "tomas.bjerre85@gmail.com",
              "authorName": "Tomas Bjerre",
              "commitTime": "2016-04-12 16:43:21",
              "commitTimeLong": 1460479401000,
              "hash": "27642e3db66e67c",
              "hashFull": "27642e3db66e67cfe17cdbcc2d6dbf8a3bbcf644",
              "merge": false,
              "message": "JENKINS-34155 Support short SHA"
            }
          ],
          "authors": [
            {
              "commits": [
                {
                  "authorEmailAddress": "tomas.bjerre85@gmail.com",
                  "authorName": "Tomas Bjerre",
                  "commitTime": "2016-04-12 16:43:21",
                  "commitTimeLong": 1460479401000,
                  "hash": "27642e3db66e67c",
                  "hashFull": "27642e3db66e67cfe17cdbcc2d6dbf8a3bbcf644",
                  "merge": false,
                  "message": "JENKINS-34155 Support short SHA"
                }
              ],
              "authorName": "Tomas Bjerre",
              "authorEmail": "tomas.bjerre85@gmail.com"
            }
          ],
          "name": "Jira",
          "title": "",
          "hasTitle": false,
          "issue": "JENKINS-34155",
          "hasIssue": true,
          "link": "",
          "hasLink": false,
          "type": "",
          "hasType": false,
          "hasDescription": false,
          "description": "",
          "hasLabels": false,
          "hasLinkedIssues": false,
          "issueType": "JIRA"
        },
        {
          "commits": [
            {
              "authorEmailAddress": "tomas.bjerre85@gmail.com",
              "authorName": "Tomas Bjerre",
              "commitTime": "2016-04-13 18:30:26",
              "commitTimeLong": 1460572226000,
              "hash": "e3106df640b693b",
              "hashFull": "e3106df640b693b50c297154919cb5564895c08e",
              "merge": false,
              "message": "Including correct commits + performance #26 JENKINS-34156\n\n * Found major performance problem when sorting tags by commit time, fixed.\n * Now not following parents, unless *from* is merged into them."
            }
          ],
          "authors": [
            {
              "commits": [
                {
                  "authorEmailAddress": "tomas.bjerre85@gmail.com",
                  "authorName": "Tomas Bjerre",
                  "commitTime": "2016-04-13 18:30:26",
                  "commitTimeLong": 1460572226000,
                  "hash": "e3106df640b693b",
                  "hashFull": "e3106df640b693b50c297154919cb5564895c08e",
                  "merge": false,
                  "message": "Including correct commits + performance #26 JENKINS-34156\n\n * Found major performance problem when sorting tags by commit time, fixed.\n * Now not following parents, unless *from* is merged into them."
                }
              ],
              "authorName": "Tomas Bjerre",
              "authorEmail": "tomas.bjerre85@gmail.com"
            }
          ],
          "name": "Jira",
          "title": "",
          "hasTitle": false,
          "issue": "JENKINS-34156",
          "hasIssue": true,
          "link": "",
          "hasLink": false,
          "type": "",
          "hasType": false,
          "hasDescription": false,
          "description": "",
          "hasLabels": false,
          "hasLinkedIssues": false,
          "issueType": "JIRA"
        }
      ],
      "issueTypes": [
        {
          "name": "Jira",
          "issues": [
            {
              "commits": [
                {
                  "authorEmailAddress": "tomas.bjerre85@gmail.com",
                  "authorName": "Tomas Bjerre",
                  "commitTime": "2016-04-12 16:43:21",
                  "commitTimeLong": 1460479401000,
                  "hash": "27642e3db66e67c",
                  "hashFull": "27642e3db66e67cfe17cdbcc2d6dbf8a3bbcf644",
                  "merge": false,
                  "message": "JENKINS-34155 Support short SHA"
                }
              ],
              "authors": [
                {
                  "commits": [
                    {
                      "authorEmailAddress": "tomas.bjerre85@gmail.com",
                      "authorName": "Tomas Bjerre",
                      "commitTime": "2016-04-12 16:43:21",
                      "commitTimeLong": 1460479401000,
                      "hash": "27642e3db66e67c",
                      "hashFull": "27642e3db66e67cfe17cdbcc2d6dbf8a3bbcf644",
                      "merge": false,
                      "message": "JENKINS-34155 Support short SHA"
                    }
                  ],
                  "authorName": "Tomas Bjerre",
                  "authorEmail": "tomas.bjerre85@gmail.com"
                }
              ],
              "name": "Jira",
              "title": "",
              "hasTitle": false,
              "issue": "JENKINS-34155",
              "hasIssue": true,
              "link": "",
              "hasLink": false,
              "type": "",
              "hasType": false,
              "hasDescription": false,
              "description": "",
              "hasLabels": false,
              "hasLinkedIssues": false,
              "issueType": "JIRA"
            },
            {
              "commits": [
                {
                  "authorEmailAddress": "tomas.bjerre85@gmail.com",
                  "authorName": "Tomas Bjerre",
                  "commitTime": "2016-04-13 18:30:26",
                  "commitTimeLong": 1460572226000,
                  "hash": "e3106df640b693b",
                  "hashFull": "e3106df640b693b50c297154919cb5564895c08e",
                  "merge": false,
                  "message": "Including correct commits + performance #26 JENKINS-34156\n\n * Found major performance problem when sorting tags by commit time, fixed.\n * Now not following parents, unless *from* is merged into them."
                }
              ],
              "authors": [
                {
                  "commits": [
                    {
                      "authorEmailAddress": "tomas.bjerre85@gmail.com",
                      "authorName": "Tomas Bjerre",
                      "commitTime": "2016-04-13 18:30:26",
                      "commitTimeLong": 1460572226000,
                      "hash": "e3106df640b693b",
                      "hashFull": "e3106df640b693b50c297154919cb5564895c08e",
                      "merge": false,
                      "message": "Including correct commits + performance #26 JENKINS-34156\n\n * Found major performance problem when sorting tags by commit time, fixed.\n * Now not following parents, unless *from* is merged into them."
                    }
                  ],
                  "authorName": "Tomas Bjerre",
                  "authorEmail": "tomas.bjerre85@gmail.com"
                }
              ],
              "name": "Jira",
              "title": "",
              "hasTitle": false,
              "issue": "JENKINS-34156",
              "hasIssue": true,
              "link": "",
              "hasLink": false,
              "type": "",
              "hasType": false,
              "hasDescription": false,
              "description": "",
              "hasLabels": false,
              "hasLinkedIssues": false,
              "issueType": "JIRA"
            }
          ],
          "type": "JIRA"
        }
      ],
      "name": "1.71",
      "tagTime": "2017-07-25 18:24:16",
      "tagTimeLong": 1501007056000,
      "hasTagTime": true
    }
  ],
  "authors": [
    {
      "commits": [
        {
          "authorEmailAddress": "tomas.bjerre85@gmail.com",
          "authorName": "Tomas Bjerre",
          "commitTime": "2016-04-13 18:30:26",
          "commitTimeLong": 1460572226000,
          "hash": "e3106df640b693b",
          "hashFull": "e3106df640b693b50c297154919cb5564895c08e",
          "merge": false,
          "message": "Including correct commits + performance #26 JENKINS-34156\n\n * Found major performance problem when sorting tags by commit time, fixed.\n * Now not following parents, unless *from* is merged into them."
        },
        {
          "authorEmailAddress": "tomas.bjerre85@gmail.com",
          "authorName": "Tomas Bjerre",
          "commitTime": "2016-04-12 16:43:21",
          "commitTimeLong": 1460479401000,
          "hash": "27642e3db66e67c",
          "hashFull": "27642e3db66e67cfe17cdbcc2d6dbf8a3bbcf644",
          "merge": false,
          "message": "JENKINS-34155 Support short SHA"
        }
      ],
      "authorName": "Tomas Bjerre",
      "authorEmail": "tomas.bjerre85@gmail.com"
    }
  ],
  "issues": [
    {
      "commits": [
        {
          "authorEmailAddress": "tomas.bjerre85@gmail.com",
          "authorName": "Tomas Bjerre",
          "commitTime": "2016-04-12 16:43:21",
          "commitTimeLong": 1460479401000,
          "hash": "27642e3db66e67c",
          "hashFull": "27642e3db66e67cfe17cdbcc2d6dbf8a3bbcf644",
          "merge": false,
          "message": "JENKINS-34155 Support short SHA"
        }
      ],
      "authors": [
        {
          "commits": [
            {
              "authorEmailAddress": "tomas.bjerre85@gmail.com",
              "authorName": "Tomas Bjerre",
              "commitTime": "2016-04-12 16:43:21",
              "commitTimeLong": 1460479401000,
              "hash": "27642e3db66e67c",
              "hashFull": "27642e3db66e67cfe17cdbcc2d6dbf8a3bbcf644",
              "merge": false,
              "message": "JENKINS-34155 Support short SHA"
            }
          ],
          "authorName": "Tomas Bjerre",
          "authorEmail": "tomas.bjerre85@gmail.com"
        }
      ],
      "name": "Jira",
      "title": "",
      "hasTitle": false,
      "issue": "JENKINS-34155",
      "hasIssue": true,
      "link": "",
      "hasLink": false,
      "type": "",
      "hasType": false,
      "hasDescription": false,
      "description": "",
      "hasLabels": false,
      "hasLinkedIssues": false,
      "issueType": "JIRA"
    },
    {
      "commits": [
        {
          "authorEmailAddress": "tomas.bjerre85@gmail.com",
          "authorName": "Tomas Bjerre",
          "commitTime": "2016-04-13 18:30:26",
          "commitTimeLong": 1460572226000,
          "hash": "e3106df640b693b",
          "hashFull": "e3106df640b693b50c297154919cb5564895c08e",
          "merge": false,
          "message": "Including correct commits + performance #26 JENKINS-34156\n\n * Found major performance problem when sorting tags by commit time, fixed.\n * Now not following parents, unless *from* is merged into them."
        }
      ],
      "authors": [
        {
          "commits": [
            {
              "authorEmailAddress": "tomas.bjerre85@gmail.com",
              "authorName": "Tomas Bjerre",
              "commitTime": "2016-04-13 18:30:26",
              "commitTimeLong": 1460572226000,
              "hash": "e3106df640b693b",
              "hashFull": "e3106df640b693b50c297154919cb5564895c08e",
              "merge": false,
              "message": "Including correct commits + performance #26 JENKINS-34156\n\n * Found major performance problem when sorting tags by commit time, fixed.\n * Now not following parents, unless *from* is merged into them."
            }
          ],
          "authorName": "Tomas Bjerre",
          "authorEmail": "tomas.bjerre85@gmail.com"
        }
      ],
      "name": "Jira",
      "title": "",
      "hasTitle": false,
      "issue": "JENKINS-34156",
      "hasIssue": true,
      "link": "",
      "hasLink": false,
      "type": "",
      "hasType": false,
      "hasDescription": false,
      "description": "",
      "hasLabels": false,
      "hasLinkedIssues": false,
      "issueType": "JIRA"
    }
  ],
  "issueTypes": [
    {
      "name": "Jira",
      "issues": [
        {
          "commits": [
            {
              "authorEmailAddress": "tomas.bjerre85@gmail.com",
              "authorName": "Tomas Bjerre",
              "commitTime": "2016-04-12 16:43:21",
              "commitTimeLong": 1460479401000,
              "hash": "27642e3db66e67c",
              "hashFull": "27642e3db66e67cfe17cdbcc2d6dbf8a3bbcf644",
              "merge": false,
              "message": "JENKINS-34155 Support short SHA"
            }
          ],
          "authors": [
            {
              "commits": [
                {
                  "authorEmailAddress": "tomas.bjerre85@gmail.com",
                  "authorName": "Tomas Bjerre",
                  "commitTime": "2016-04-12 16:43:21",
                  "commitTimeLong": 1460479401000,
                  "hash": "27642e3db66e67c",
                  "hashFull": "27642e3db66e67cfe17cdbcc2d6dbf8a3bbcf644",
                  "merge": false,
                  "message": "JENKINS-34155 Support short SHA"
                }
              ],
              "authorName": "Tomas Bjerre",
              "authorEmail": "tomas.bjerre85@gmail.com"
            }
          ],
          "name": "Jira",
          "title": "",
          "hasTitle": false,
          "issue": "JENKINS-34155",
          "hasIssue": true,
          "link": "",
          "hasLink": false,
          "type": "",
          "hasType": false,
          "hasDescription": false,
          "description": "",
          "hasLabels": false,
          "hasLinkedIssues": false,
          "issueType": "JIRA"
        },
        {
          "commits": [
            {
              "authorEmailAddress": "tomas.bjerre85@gmail.com",
              "authorName": "Tomas Bjerre",
              "commitTime": "2016-04-13 18:30:26",
              "commitTimeLong": 1460572226000,
              "hash": "e3106df640b693b",
              "hashFull": "e3106df640b693b50c297154919cb5564895c08e",
              "merge": false,
              "message": "Including correct commits + performance #26 JENKINS-34156\n\n * Found major performance problem when sorting tags by commit time, fixed.\n * Now not following parents, unless *from* is merged into them."
            }
          ],
          "authors": [
            {
              "commits": [
                {
                  "authorEmailAddress": "tomas.bjerre85@gmail.com",
                  "authorName": "Tomas Bjerre",
                  "commitTime": "2016-04-13 18:30:26",
                  "commitTimeLong": 1460572226000,
                  "hash": "e3106df640b693b",
                  "hashFull": "e3106df640b693b50c297154919cb5564895c08e",
                  "merge": false,
                  "message": "Including correct commits + performance #26 JENKINS-34156\n\n * Found major performance problem when sorting tags by commit time, fixed.\n * Now not following parents, unless *from* is merged into them."
                }
              ],
              "authorName": "Tomas Bjerre",
              "authorEmail": "tomas.bjerre85@gmail.com"
            }
          ],
          "name": "Jira",
          "title": "",
          "hasTitle": false,
          "issue": "JENKINS-34156",
          "hasIssue": true,
          "link": "",
          "hasLink": false,
          "type": "",
          "hasType": false,
          "hasDescription": false,
          "description": "",
          "hasLabels": false,
          "hasLinkedIssues": false,
          "issueType": "JIRA"
        }
      ],
      "type": "JIRA"
    }
  ],
  "ownerName": "tomasbjerre",
  "repoName": "git-changelog-lib"
}